PR for #170: let the window option choose a window range

Constantino-Carreto-Romero opened the pull request:
Closes #170

jorpppp commented on xtevent/xtevent.ado:
```diff
@@ -348,6 +374,67 @@ program define xtevent, eclass
+		
+		if `nwwindow'==1 {
+			if !inlist("`anything'","max","balanced"){
+				di as err _n "{bf:window} must be one of {bf:max} or {bf:balanced}."
```
@Constantino-Carreto-Romero  "must be max or balanced"

Constantino-Carreto-Romero replied:
@jorpppp 
changed in https://github.com/JMSLab/xtevent/commit/4ee9081ff734c9efaaaaadc86b0033cee895ae22

jorpppp commented on xtevent/_eventgenvars.ado:
```diff
@@ -220,6 +224,12 @@ program define _eventgenvars, rclass
+	
+	if `bin'==0 & "`w_type'"=="string" {
+		di "The policy variable is not binary. Assuming non-staggered adoption."
+		di as err _n "Option {bf:window(`lwindow')} can be used only if policyvar follows staggered adoption."
```
@Constantino-Carreto-Romero this is only true with a non-binary policy variable. With a binary policy variable window can be used with impute(nuchange) even without staggered adoption, right? Modify the error message accordingly.

Constantino-Carreto-Romero replied:
@jorpppp thanks
The error message will be one of:
- Option `window(max)` can be used only if policyvar follows staggered adoption
-  Option `window(balanced)` can be used only if policyvar follows staggered adoption

because at this point in `_eventgenvars` `lwindow` (and `rwindow`) equals `max` or `balanced` if the local `w_type` equals `string`. 
So I think the error message is fine, but let me know if I'm missing something.

jorpppp replied:
@Constantino-Carreto-Romero The bigger question is, why can't you use `window(max)` or `window(balanced)` outside of staggered adoption?

jorpppp replied:
@Constantino-Carreto-Romero We will only enable these options under staggered adoption so the message is fine.

jorpppp commented:
@Constantino-Carreto-Romero If I run
`xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max)`
with the test data I get a no observations error. This is counterintuive since 
`xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max)` 
does run. So I would think the maximum window in this setting is at least 5. There's no imputation here, is there a bug when calculating the max window without impute?

jorpppp commented on xtevent/_eventgenvars.ado:
```diff
@@ -173,7 +177,7 @@ program define _eventgenvars, rclass
 	loc bounds 0
 	* show a warning message if we don't know treatment time for some units due to missing values in policyvar 
-	if ("`impute'"=="stag" | "`impute'"=="instag") {
+	if ("`impute'"=="stag" | "`impute'"=="instag" | "`w_type'"=="string") {
```
@Constantino-Carreto-Romero I understand that this piece of code is because you want the check below to be ran if window is max or balanced. But this is confusing because these are not excluding cases, right? You can have impute be stag AND w_type can be a string.

Is there a better way to structure this check such that the condition here is only over the impute cases, and instead you run the checks in another place if w_type is a string?

Constantino-Carreto-Romero replied:
@jorpppp 
in https://github.com/JMSLab/xtevent/commit/caec36a0c8b136fe639c163a86fdf2f944ff76a3 I restructured the checks for staggered adoption and now they happend only if `impute(stag | instag)` is specified. Requirement of the `impute` option was implemented in https://github.com/JMSLab/xtevent/pull/178#issuecomment-2088869630. If policyvar doesn't follow staggered adoption, first, there will be a warning message saying why the policyvar doesn't follow staggered adoption. Then, an error message will be issue saying that `window(max | balanced)` cannot be used if policyvar doesn't follow staggered adoption.
For instance:
```
*it reverts
use example31, clear
replace z= 0 in 8
cap noi xtevent y eta , panelvar(i) timevar(t) policyvar(z) impute(stag) window(max)
```
generates the following [message](https://github.com/JMSLab/xtevent/assets/36095845/355f5bf8-31b1-4241-9540-e2f04f6b0f6d).

jorpppp commented on xtevent/_eventiv.ado:
```diff
@@ -284,11 +295,16 @@ program define _eventiv, rclass
 	
```
@Constantino-Carreto-Romero  Do we need a check for whether the proxyiv lags to be normalized are in the estimation window if window was chosen to be max or balanced?

Constantino-Carreto-Romero replied:
@jorpppp 
For IV estimation, `_eventiv` makes two calls to `_eventgenvars`. The first call is to bring the imputed policyvar because it uses it to choose the lead order for the instrument. `_eventiv` also benefits from this first call and brings the calculated left window in https://github.com/JMSLab/xtevent/blob/267cef77ee4a32a1c26cce9d41e129d88cd45be0/xtevent/_eventiv.ado#L80. Then, in https://github.com/JMSLab/xtevent/blob/267cef77ee4a32a1c26cce9d41e129d88cd45be0/xtevent/_eventiv.ado#L86 it uses that value to define the limit for the lead orders to check in the code that selects the lead order later on in https://github.com/JMSLab/xtevent/blob/267cef77ee4a32a1c26cce9d41e129d88cd45be0/xtevent/_eventiv.ado#L135.

jorpppp replied:
@Constantino-Carreto-Romero So, if the user chose `window(max)` then first the window will be determined, and then the lead order will be chosen in that window. Sounds good.

However, my comment above was for the case when the user selects the leads, e.g. `window(max) proxyiv(2 3)`. Is that case handled by the second call to `event_iv`, such that it will check if the leads are in the chose window?

Constantino-Carreto-Romero replied:
@jorpppp 
There didn't exist such a check, even for the numeric window case. In https://github.com/JMSLab/xtevent/commit/74fa61b1888eb9165349e2ff7f69ba0b766623a6 I implemented such a check which works for both the numeric-defined window or the max/balanced case. 
For instance, when trying to run:
```
use example31, clear
cap noi xtevent y eta, panelvar(i) timevar(t) policyvar(z) proxy(x) impute(stag) window(max) proxyiv(5 e 20)

```
shows the following [error message](https://github.com/JMSLab/xtevent/assets/36095845/ca6560fe-fd37-4e58-8acd-404ce10ee86e).

jorpppp replied:
@Constantino-Carreto-Romero looks good, thanks.

jorpppp commented on xtevent/xteventplot.ado:
```diff
@@ -622,7 +622,7 @@ end
 program define parsecmdline, rclass
-	syntax anything [aw fw pw] [if][in], samplevar(string) [Window(numlist min=1 max=2 integer) savek(string) plot proxy(string) POLicyvar(string) impute(string) *]
+	syntax anything [aw fw pw] [if][in], samplevar(string) [Window(string) savek(string) plot proxy(string) POLicyvar(string) impute(string) *]
 	
```
@Constantino-Carreto-Romero Just confirming something. We don't need to include the window option parser inside xteventplot here because the window option is already "valid", in the sense that it was already parsed and passed checks when ran by xtevent. Is that so?

Constantino-Carreto-Romero replied:
@jorpppp 
Yes, if `xtevent` completed execution, `window(max | balanced)` passed all checks and is valid. When executing `xteventplot` and parsing `cmdline` for overlay plots, for instance for the overlay static plot `xteventplot` will call `xtevent` in https://github.com/JMSLab/xtevent/blob/267cef77ee4a32a1c26cce9d41e129d88cd45be0/xtevent/xteventplot.ado#L252 using the original specification of `window`: `window(max)` or `window(balanced)`. It will be passed to `xtevent` along with the `impute` option because it was specified and passed checks in the original call to `xtevent`. Therefore, there are not inconveniences with `xteventplot` passing to `xtevent` the original specification of `window`.
The following example covers that case and runs without inconveniences
```
use example31, clear
xtevent y eta , panelvar(i) timevar(t) policyvar(z) impute(stag) window(max)
xteventplot, overlay(static)
```

jorpppp commented on xtevent/xtevent.sthlp:
```diff
@@ -98,12 +98,13 @@ have not been previously {cmd:xtset}. See {help xtset}.
 estimation will use an asymmetric window with {it:k1} periods before the event and {it:k2} periods after the event. For example, with {it:k1} = -1 
-and {it:k2} = 2, there will be four coefficients in the window (-1,0,1,2) and two endpoints (-2+,3+). {opt window()} is required unless 
-{opt static} is specified, or if the estimation window is specified using  options {opt pre()}, {opt post()}, {opt overidpre()}, 
+and {it:k2} = 2, there will be four coefficients in the window (-1,0,1,2) and two endpoints (-2+,3+). Notice that if the event time in your database goes from -2 to 3, the maximum window you can specify is then {it:k1} = -1 and 
```
@Constantino-Carreto-Romero Can we split the explanation of the numeric window options and the max, balanced options in different paragraphs?

Constantino-Carreto-Romero replied:
@jorpppp 
Changed in https://github.com/JMSLab/xtevent/commit/a4002b575e8c39fad8c8342d4366ba48e8aad6be

jorpppp commented:
@Constantino-Carreto-Romero I get an error when running
`xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max) impute(stag) trend(-2)`
This happens because the check for whether the trend starting point is in the estimation window is using the values of `lwindow` and `rwindow` and at the time of the check these are strings. Maybe the check needs to be later, and it must use the calculated window limits.

jorpppp commented on xtevent/_eventgenvars.ado:
```diff
@@ -310,6 +332,64 @@ program define _eventgenvars, rclass
+		loc rwindow = `rwindow' -1
+		
+		**** Error messages if calculated window limits are not valid  
+		if  (-`lwindow'<0 | `rwindow'<0) {
```
@Constantino-Carreto-Romero The message displaying the calculated window limits below should appear even if there are no errors. It will be informative to know what are the window limits in all cases.
Also add an informative message on which window selection criteria was used.

Constantino-Carreto-Romero replied:
@jorpppp 
changed in https://github.com/JMSLab/xtevent/commit/267cef77ee4a32a1c26cce9d41e129d88cd45be0. 
I also changed an error message to be more informative.

jorpppp replied:
@Constantino-Carreto-Romero What does "around the event-time" mean? I think you mean the time when these first treated and last treated units are treated, right? If so change the message because we use "event-time" to refer to time to event in general.

Constantino-Carreto-Romero replied:
@jorpppp 
changed in https://github.com/JMSLab/xtevent/commit/154d09f0c66566bf6c41a6841031c0a4de895964

jorpppp commented:
@Constantino-Carreto-Romero If I run 
`xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(balanced) impute(stag)`
I get an error message. The found window limits are -1 and -1.

This doesn't sem right. If we run
```
xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(balanced) impute(stag) savek(event)
bys i: egen min = min(event_evtime)
bys i: egen max = max(event_evtime)
preserve
bys i: keep if _n == 1
tab min
tab max
restore 
```

We get that all treated units have at least two periods pretreatment and one period post treatment (the treatment period, 0). So the balanced window should be -2, 0. Why is it not the case? Am I misreading how the balanced window should be calculated?

jorpppp commented:
@Constantino-Carreto-Romero please run more checks interacting the new window options and the other options in `xtevent`. The fact that I was able to find clashes when using `trend` or using the balanced option makes me think that we need more checking before we are ready to merge.

jorpppp commented:
@Constantino-Carreto-Romero  Note changes in #181 to avoid conflicts when merging this PR

Constantino-Carreto-Romero commented:
> @Constantino-Carreto-Romero If I run `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max)` with the test data I get a no observations error. This is counterintuive since `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(5)` does run. So I would think the maximum window in this setting is at least 5. There's no imputation here, is there a bug when calculating the max window without impute?

@jorpppp thanks
If I run the following example
```
use example31.dta, clear
xtset i t 
*generate standardized time variable 
gen diffpolicyvar=d1.z
gen timeoftreatment=t if diffpolicyvar==1
by i: egen ttreat= min(timeoftreatment)
gen t_relative = t - ttreat
*tabulate the standardized time variable
tab t_relative
```
when tabulating the standardized time variable I get the [following](https://github.com/JMSLab/xtevent/assets/36095845/f77926a5-491b-4757-ac76-600eeb42bb62). It indicates that the maximum range of estimation is -19, 17. But then, adjusting for the endpoints, the maximum possible window the user could specify is `window(-18 16)`. Internally, the program tries to estimate using that window. Nonetheless, since `impute` is not active, the event-time dummies are generated with missing values. For instance, running the following
```
xtevent y, panelvar(i) timevar(t) policyvar(z) w(-18 16) savek(aa, noestimate) 
```
and then taking a look at the event-time dummies, or checking 
```
count if !missing(aa_eq_m19) & !missing(aa_eq_p17)
```
it returns 0. Therefore, no observation succeeds in having non-missing values for all the event-time dummies. 
Now if I use the `eventdd` command, I get the following [result](https://github.com/JMSLab/xtevent/assets/36095845/5c6872f7-8fe0-425d-9160-4111d4ec55ce). By default, `eventdd` estimates the maximum possible window, in this case, 19 leads and 17 lags (`eventdd` does not differentiate the endpoints from the rest of event-time dummies). `eventdd` succeeds at estimating because by default it assumes staggered adoption and therefore it imputes the event-time dummies and endpoints. If I now try `xtevent` with the `impute` option:
```
xtevent y, panelvar(i) timevar(t) policyvar(z) w(-18 16) impute(stag) 
```
I get the same estimation [range](https://github.com/JMSLab/xtevent/assets/36095845/f757d66a-7f09-4b5e-a04e-0f37081a8858) as with `eventdd`. Therefore, in the case of `xtevent`, with the default of not activating the `impute` option, success in the estimation will depend on how many event-time dummies coefficients it has to estimate, according to the calculated window. If the dataset has a short observed time range, the maximum possible window will likely be short enough such that even without the `impute` option, the program will complete the estimation.

jorpppp commented:
> > @Constantino-Carreto-Romero If I run `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max)` with the test data I get a no observations error. This is counterintuive since `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(5)` does run. So I would think the maximum window in this setting is at least 5. There's no imputation here, is there a bug when calculating the max window without impute?
> 
> @jorpppp thanks If I run the following example
> 
> ```
> use example31.dta, clear
> xtset i t 
> *generate standardized time variable 
> gen diffpolicyvar=d1.z
> gen timeoftreatment=t if diffpolicyvar==1
> by i: egen ttreat= min(timeoftreatment)
> gen t_relative = t - ttreat
> *tabulate the standardized time variable
> tab t_relative
> ```
> 
> when tabulating the standardized time variable I get the [following](https://github.com/JMSLab/xtevent/assets/36095845/f77926a5-491b-4757-ac76-600eeb42bb62). It indicates that the maximum range of estimation is -19, 17. But then, adjusting for the endpoints, the maximum possible window the user could specify is `window(-18 16)`. Internally, the program tries to estimate using that window. Nonetheless, since `impute` is not active, the event-time dummies are generated with missing values. For instance, running the following
> 
> ```
> xtevent y, panelvar(i) timevar(t) policyvar(z) w(-18 16) savek(aa, noestimate) 
> ```
> 
> and then taking a look at the event-time dummies, or checking
> 
> ```
> count if !missing(aa_eq_m19) & !missing(aa_eq_p17)
> ```
> 
> it returns 0. Therefore, no observation succeeds in having non-missing values for all the event-time dummies. Now if I use the `eventdd` command, I get the following [result](https://github.com/JMSLab/xtevent/assets/36095845/5c6872f7-8fe0-425d-9160-4111d4ec55ce). By default, `eventdd` estimates the maximum possible window, in this case, 19 leads and 17 lags (`eventdd` does not differentiate the endpoints from the rest of event-time dummies). `eventdd` succeeds at estimating because by default it assumes staggered adoption and therefore it imputes the event-time dummies and endpoints. If I now try `xtevent` with the `impute` option:
> 
> ```
> xtevent y, panelvar(i) timevar(t) policyvar(z) w(-18 16) impute(stag) 
> ```
> 
> I get the same estimation [range](https://github.com/JMSLab/xtevent/assets/36095845/f757d66a-7f09-4b5e-a04e-0f37081a8858) as with `eventdd`. Therefore, in the case of `xtevent`, with the default of not activating the `impute` option, success in the estimation will depend on how many event-time dummies coefficients it has to estimate, according to the calculated window. If the dataset has a short observed time range, the maximum possible window will likely be short enough such that even without the `impute` option, the program will complete the estimation.

@Constantino-Carreto-Romero What I am reading from this is that the criteria to select the maximum window that we are using only works for staggered adoption, and that we need to figure out a criterion that works outside the staggered adoption case. Can you figure out what the criterion is?

jorpppp commented:
> > > @Constantino-Carreto-Romero If I run `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max)` with the test data I get a no observations error. This is counterintuive since `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(5)` does run. So I would think the maximum window in this setting is at least 5. There's no imputation here, is there a bug when calculating the max window without impute?
> > 
> > 
> > @jorpppp thanks If I run the following example
> > ```
> > use example31.dta, clear
> > xtset i t 
> > *generate standardized time variable 
> > gen diffpolicyvar=d1.z
> > gen timeoftreatment=t if diffpolicyvar==1
> > by i: egen ttreat= min(timeoftreatment)
> > gen t_relative = t - ttreat
> > *tabulate the standardized time variable
> > tab t_relative
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > when tabulating the standardized time variable I get the [following](https://github.com/JMSLab/xtevent/assets/36095845/f77926a5-491b-4757-ac76-600eeb42bb62). It indicates that the maximum range of estimation is -19, 17. But then, adjusting for the endpoints, the maximum possible window the user could specify is `window(-18 16)`. Internally, the program tries to estimate using that window. Nonetheless, since `impute` is not active, the event-time dummies are generated with missing values. For instance, running the following
> > ```
> > xtevent y, panelvar(i) timevar(t) policyvar(z) w(-18 16) savek(aa, noestimate) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and then taking a look at the event-time dummies, or checking
> > ```
> > count if !missing(aa_eq_m19) & !missing(aa_eq_p17)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it returns 0. Therefore, no observation succeeds in having non-missing values for all the event-time dummies. Now if I use the `eventdd` command, I get the following [result](https://github.com/JMSLab/xtevent/assets/36095845/5c6872f7-8fe0-425d-9160-4111d4ec55ce). By default, `eventdd` estimates the maximum possible window, in this case, 19 leads and 17 lags (`eventdd` does not differentiate the endpoints from the rest of event-time dummies). `eventdd` succeeds at estimating because by default it assumes staggered adoption and therefore it imputes the event-time dummies and endpoints. If I now try `xtevent` with the `impute` option:
> > ```
> > xtevent y, panelvar(i) timevar(t) policyvar(z) w(-18 16) impute(stag) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get the same estimation [range](https://github.com/JMSLab/xtevent/assets/36095845/f757d66a-7f09-4b5e-a04e-0f37081a8858) as with `eventdd`. Therefore, in the case of `xtevent`, with the default of not activating the `impute` option, success in the estimation will depend on how many event-time dummies coefficients it has to estimate, according to the calculated window. If the dataset has a short observed time range, the maximum possible window will likely be short enough such that even without the `impute` option, the program will complete the estimation.
> 
> @Constantino-Carreto-Romero What I am reading from this is that the criteria to select the maximum window that we are using only works for staggered adoption, and that we need to figure out a criterion that works outside the staggered adoption case. Can you figure out what the criterion is?

Per call: we won't try to deal with the case of automatic window choice outside staggered adoption. @Constantino-Carreto-Romero

jorpppp commented:
@Constantino-Carreto-Romero we won't deal with the case outside staggered adoption. So the automatic window generation commands should check for staggered adoption before proceding, and return an error otherwise.

The other open comments in this PR are also pending.

jorpppp commented:
@Constantino-Carreto-Romero Keep the changes in #187 in mind when merging this PR

Constantino-Carreto-Romero commented:
> @Constantino-Carreto-Romero I get an error when running `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max) impute(stag) trend(-2)` This happens because the check for whether the trend starting point is in the estimation window is using the values of `lwindow` and `rwindow` and at the time of the check these are strings. Maybe the check needs to be later, and it must use the calculated window limits.

@jorpppp 
In https://github.com/JMSLab/xtevent/commit/de25d54a0c276cce974e52e8fce47720d166f612 and https://github.com/JMSLab/xtevent/commit/9fcf9b35c789fbda477e942821b8195d2df2b2f4 I kept the check in `_eventols` but conditioning that window must be numeric. Then, in `_eventgenvars`, once the window is calculated (if `window(max | balanced)` was specified), the same check is carried out. The check must be placed in `_eventgenvars`, before the code that specifies the coefficients that will be normalized.
I kept the check in `_eventols` for the numeric window case, so the program can stop at that instance and not let it stop until `_eventgenvars`. Should I change it to a single check in `_eventgenvars`?

jorpppp commented:
> > @Constantino-Carreto-Romero I get an error when running `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(max) impute(stag) trend(-2)` This happens because the check for whether the trend starting point is in the estimation window is using the values of `lwindow` and `rwindow` and at the time of the check these are strings. Maybe the check needs to be later, and it must use the calculated window limits.
> 
> @jorpppp In [de25d54](https://github.com/JMSLab/xtevent/commit/de25d54a0c276cce974e52e8fce47720d166f612) and [9fcf9b3](https://github.com/JMSLab/xtevent/commit/9fcf9b35c789fbda477e942821b8195d2df2b2f4) I kept the check in `_eventols` but conditioning that window must be numeric. Then, in `_eventgenvars`, once the window is calculated (if `window(max | balanced)` was specified), the same check is carried out. The check must be placed in `_eventgenvars`, before the code that specifies the coefficients that will be normalized. I kept the check in `_eventols` for the numeric window case, so the program can stop at that instance and not let it stop until `_eventgenvars`. Should I change it to a single check in `_eventgenvars`?

@Constantino-Carreto-Romero Thanks. It's ok to keep the check in both places.

Constantino-Carreto-Romero commented:
> > > @Constantino-Carreto-Romero If I run `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(balanced) impute(stag)` I get an error message. The found window limits are -1 and -1.
> > > This doesn't seem right. If we run
> > > ```
> > > xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(balanced) impute(stag) savek(event)
> > > bys i: egen min = min(event_evtime)
> > > bys i: egen max = max(event_evtime)
> > > preserve
> > > bys i: keep if _n == 1
> > > tab min
> > > tab max
> > > restore 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > We get that all treated units have at least two periods pretreatment and one period post treatment (the treatment period, 0). So the balanced window should be -2, 0. Why is it not the case? Am I misreading how the balanced window should be calculated?
> > 
> > 
> > @jorpppp The calculated window is (-1 -1) plus the endpoints (-2+, 0+). The calculated limits are indeed (-2 0) but I follow the definition used for a numeric input for `window` and separate the periods covered by `window` and the endpoints. After the changes in [267cef7](https://github.com/JMSLab/xtevent/commit/267cef77ee4a32a1c26cce9d41e129d88cd45be0), running the same example produces this more informative [error message](https://github.com/JMSLab/xtevent/assets/36095845/a553cc92-ce01-4653-be42-ec5a57c5a912).
> 
> @Constantino-Carreto-Romero Thanks. The error message is still a bit confusing. Maybe write "the calculated window is (-1,-1), excluding the endpoints". The -2+, 0+ is confusing, I don't know what the plus signs mean. So I would omit it.
> 


@jorpppp 
I edited the message in https://github.com/JMSLab/xtevent/commit/8409db79fc408499f779215ade1eb9fe0f43f28b. Is it ok?


> In testing this I found a possible bug: `xtevent y, w(0 3) pol(z) impute(stag)` returns an error with the example31 data. Can you check please?


I tested it and I got the error message. The error comes from this line https://github.com/JMSLab/xtevent/blob/8409db79fc408499f779215ade1eb9fe0f43f28b/xtevent/_eventgenvars.ado#L454 where the temporary variable being analyzed was not defined because this code section https://github.com/JMSLab/xtevent/blob/8409db79fc408499f779215ade1eb9fe0f43f28b/xtevent/_eventgenvars.ado#L430 was not executed because the left window is 0. I'm not sure whether it is a bug. It is not related to the automatic-window implementation, can I open a separate issue to check it?

jorpppp commented:
> > > > @Constantino-Carreto-Romero If I run `xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(balanced) impute(stag)` I get an error message. The found window limits are -1 and -1.
> > > > This doesn't seem right. If we run
> > > > ```
> > > > xtevent y eta, panelvar(i) timevar(t) policyvar(z) w(balanced) impute(stag) savek(event)
> > > > bys i: egen min = min(event_evtime)
> > > > bys i: egen max = max(event_evtime)
> > > > preserve
> > > > bys i: keep if _n == 1
> > > > tab min
> > > > tab max
> > > > restore 
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > We get that all treated units have at least two periods pretreatment and one period post treatment (the treatment period, 0). So the balanced window should be -2, 0. Why is it not the case? Am I misreading how the balanced window should be calculated?
> > > 
> > > 
> > > @jorpppp The calculated window is (-1 -1) plus the endpoints (-2+, 0+). The calculated limits are indeed (-2 0) but I follow the definition used for a numeric input for `window` and separate the periods covered by `window` and the endpoints. After the changes in [267cef7](https://github.com/JMSLab/xtevent/commit/267cef77ee4a32a1c26cce9d41e129d88cd45be0), running the same example produces this more informative [error message](https://github.com/JMSLab/xtevent/assets/36095845/a553cc92-ce01-4653-be42-ec5a57c5a912).
> > 
> > 
> > @Constantino-Carreto-Romero Thanks. The error message is still a bit confusing. Maybe write "the calculated window is (-1,-1), excluding the endpoints". The -2+, 0+ is confusing, I don't know what the plus signs mean. So I would omit it.
> 
> @jorpppp I edited the message in [8409db7](https://github.com/JMSLab/xtevent/commit/8409db79fc408499f779215ade1eb9fe0f43f28b). Is it ok?
> 

@Constantino-Carreto-Romero let me make a minor change to it in a push. After that I'll test the code a bit more and if it's ok I'll merge the PR.


> > In testing this I found a possible bug: `xtevent y, w(0 3) pol(z) impute(stag)` returns an error with the example31 data. Can you check please?
> 
> I tested it and I got the error message. The error comes from this line
> 
> https://github.com/JMSLab/xtevent/blob/8409db79fc408499f779215ade1eb9fe0f43f28b/xtevent/_eventgenvars.ado#L454
> 
> where the temporary variable being analyzed was not defined because this code section
> https://github.com/JMSLab/xtevent/blob/8409db79fc408499f779215ade1eb9fe0f43f28b/xtevent/_eventgenvars.ado#L430
> 
> was not executed because the left window is 0. I'm not sure whether it is a bug. It is not related to the automatic-window implementation, can I open a separate issue to check it?

@Constantino-Carreto-Romero Yes, please open a new issue for this.

jorpppp commented:
@Constantino-Carreto-Romero We still have bugs in this. With the example31 data:

```
gen pois = rpoisson(5) in 1/200
xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
```
returns an error message. I presume `touse` is not being set somewhere and the calculated window is using the entire sample and not the `touse` sample. Note that if we pass a sample restriction to `xtevent` this works:

```
xtevent y eta i.pois if pois!=., panelvar(i) timevar(t) pol(z) window(max) impute(stag)
```

Can you check this please?

jorpppp commented:
> > @Constantino-Carreto-Romero We still have bugs in this. With the example31 data:
> > ```
> > gen pois = rpoisson(5) in 1/200
> > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > returns an error message. I presume `touse` is not being set somewhere and the calculated window is using the entire sample and not the `touse` sample. Note that if we pass a sample restriction to `xtevent` this works:
> > ```
> > xtevent y eta i.pois if pois!=., panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you check this please?
> 
> @jorpppp I tried running the same example but cannot replicate the error message. When running
> 
> ```
> use example31, clear
> gen pois = rpoisson(5) in 1/200
> xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> ```
> 
> I get the following [outcome](https://github.com/JMSLab/xtevent/assets/36095845/1772439e-7a18-4da2-af63-5e2446edd903).

@Constantino-Carreto-Romero did not get the error because he wasn't trying to plot. Strictly speaking the return message appears with 
```
xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag) plot
```
In any case there is a bug because the calculated window is not taking the missing values of the control variables into account, but it should. The issue has to do with the way samples are marked throughout: `_eventgenvars` is working with a sample that has not been marked for missing values of the variables, which is convenient for other reasons, e.g. imputation. @Constantino-Carreto-Romero will check this.

Constantino-Carreto-Romero commented:
> > > @Constantino-Carreto-Romero We still have bugs in this. With the example31 data:
> > > ```
> > > gen pois = rpoisson(5) in 1/200
> > > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > returns an error message. I presume `touse` is not being set somewhere and the calculated window is using the entire sample and not the `touse` sample. Note that if we pass a sample restriction to `xtevent` this works:
> > > ```
> > > xtevent y eta i.pois if pois!=., panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Can you check this please?
> > 
> > 
> > @jorpppp I tried running the same example but cannot replicate the error message. When running
> > ```
> > use example31, clear
> > gen pois = rpoisson(5) in 1/200
> > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I get the following [outcome](https://github.com/JMSLab/xtevent/assets/36095845/1772439e-7a18-4da2-af63-5e2446edd903).
> 
> @Constantino-Carreto-Romero did not get the error because he wasn't trying to plot. Strictly speaking the error message appears with
> 
> ```
> xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag) plot
> ```
> 
> In any case there is a bug because the calculated window is not taking the missing values of the control variables into account, but it should. The issue has to do with the way samples are marked throughout: `_eventgenvars` is working with a sample that has not been marked for missing values of the variables, which is convenient for other reasons, e.g. imputation. @Constantino-Carreto-Romero will check this.

@jorpppp 
in https://github.com/JMSLab/xtevent/commit/6dcd33c45d6b77922f60bd2c02dc4d84e707fb2e I modified the syntax of `_eventgenvars`, so it can handle the local `varlist`. Then, I added a `marksample` to mark the non-missing observations in `varlist` and consider this restriction when calculating the window limits. This doesn't affect the already defined `mark` because the latter only considers `if` or `in` restrictions. I also updated the calls to `_eventgenvars` from other programs.  I include in the do-file in the issue folder examples to test that: 
- the calculation of the window limits considers missing values in the local `varlist` 
- the consideration of `if` or `in` conditions were not affected

Constantino-Carreto-Romero commented:
@jorpppp 
in https://github.com/JMSLab/xtevent/pull/178/commits/97dff856316fb4f9fcdb8abad2016b63f0a46791 I uploaded an issue folder with a do-file with examples to test: i) error messages; ii) correct functionality; iii) interaction with other options and iv; that the implementation doesn't alter other functionalities.

jorpppp commented:
> > > > @Constantino-Carreto-Romero We still have bugs in this. With the example31 data:
> > > > ```
> > > > gen pois = rpoisson(5) in 1/200
> > > > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > returns an error message. I presume `touse` is not being set somewhere and the calculated window is using the entire sample and not the `touse` sample. Note that if we pass a sample restriction to `xtevent` this works:
> > > > ```
> > > > xtevent y eta i.pois if pois!=., panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Can you check this please?
> > > 
> > > 
> > > @jorpppp I tried running the same example but cannot replicate the error message. When running
> > > ```
> > > use example31, clear
> > > gen pois = rpoisson(5) in 1/200
> > > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I get the following [outcome](https://github.com/JMSLab/xtevent/assets/36095845/1772439e-7a18-4da2-af63-5e2446edd903).
> > 
> > 
> > @Constantino-Carreto-Romero did not get the error because he wasn't trying to plot. Strictly speaking the error message appears with
> > ```
> > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag) plot
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In any case there is a bug because the calculated window is not taking the missing values of the control variables into account, but it should. The issue has to do with the way samples are marked throughout: `_eventgenvars` is working with a sample that has not been marked for missing values of the variables, which is convenient for other reasons, e.g. imputation. @Constantino-Carreto-Romero will check this.
> 
> @jorpppp in [6dcd33c](https://github.com/JMSLab/xtevent/commit/6dcd33c45d6b77922f60bd2c02dc4d84e707fb2e) I modified the syntax of `_eventgenvars`, so it can handle the local `varlist`. Then, I added a `marksample` to mark the non-missing observations in `varlist` and consider this restriction when calculating the window limits. This doesn't affect the already defined `mark` because the latter only considers `if` or `in` restrictions. I also updated the calls to `_eventgenvars` from other programs. I include in the do-file in the issue folder examples to test that:
> 
> * the calculation of the window limits considers missing values in the local `varlist`
> * the consideration of `if` or `in` conditions were not affected
> 
> examples:
> 
> https://github.com/JMSLab/xtevent/blob/97dff856316fb4f9fcdb8abad2016b63f0a46791/issue170/issue170.do#L78

Thanks @Constantino-Carreto-Romero . This approach works but has a slight disadvantage. `_eventgenvars` is recalculating the marked sample that considers missing values in the variables. But this marked sample was already calculated by `_eventols` or `_eventiv`.

So instead of passing the varlist to `_eventgenvars` to repeat the marking, we can just pass the marked sample variable from `_eventols` or `_eventiv` to eventgenvars as an option.

jorpppp commented:
> > > > @Constantino-Carreto-Romero We still have bugs in this. With the example31 data:
> > > > ```
> > > > gen pois = rpoisson(5) in 1/200
> > > > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > returns an error message. I presume `touse` is not being set somewhere and the calculated window is using the entire sample and not the `touse` sample. Note that if we pass a sample restriction to `xtevent` this works:
> > > > ```
> > > > xtevent y eta i.pois if pois!=., panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Can you check this please?
> > > 
> > > 
> > > @jorpppp I tried running the same example but cannot replicate the error message. When running
> > > ```
> > > use example31, clear
> > > gen pois = rpoisson(5) in 1/200
> > > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I get the following [outcome](https://github.com/JMSLab/xtevent/assets/36095845/1772439e-7a18-4da2-af63-5e2446edd903).
> > 
> > 
> > @Constantino-Carreto-Romero did not get the error because he wasn't trying to plot. Strictly speaking the error message appears with
> > ```
> > xtevent y eta i.pois, panelvar(i) timevar(t) pol(z) window(max) impute(stag) plot
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In any case there is a bug because the calculated window is not taking the missing values of the control variables into account, but it should. The issue has to do with the way samples are marked throughout: `_eventgenvars` is working with a sample that has not been marked for missing values of the variables, which is convenient for other reasons, e.g. imputation. @Constantino-Carreto-Romero will check this.
> 
> @jorpppp in [6dcd33c](https://github.com/JMSLab/xtevent/commit/6dcd33c45d6b77922f60bd2c02dc4d84e707fb2e) I modified the syntax of `_eventgenvars`, so it can handle the local `varlist`. Then, I added a `marksample` to mark the non-missing observations in `varlist` and consider this restriction when calculating the window limits. This doesn't affect the already defined `mark` because the latter only considers `if` or `in` restrictions. I also updated the calls to `_eventgenvars` from other programs. I include in the do-file in the issue folder examples to test that:
> 
> * the calculation of the window limits considers missing values in the local `varlist`
> * the consideration of `if` or `in` conditions were not affected
> 
> examples:
> 
> https://github.com/JMSLab/xtevent/blob/97dff856316fb4f9fcdb8abad2016b63f0a46791/issue170/issue170.do#L78

Thanks @Constantino-Carreto-Romero . This approach works but has a slight disadvantage. _eventgenvars is recalculating the marked sample that considers missing values in the variables. But this marked sample was already calculated by _eventols or _eventiv.

So instead of passing the varlist to _eventgenvars to repeat the marking, we can just pass the marked sample variable from _eventols or _eventiv to eventgenvars as an option.

Constantino-Carreto-Romero commented:
> Thanks @Constantino-Carreto-Romero . This approach works but has a slight disadvantage. `_eventgenvars` is recalculating the marked sample that considers missing values in the variables. But this marked sample was already calculated by `_eventols` or `_eventiv`.
> 
> So instead of passing the varlist to `_eventgenvars` to repeat the marking, we can just pass the marked sample variable from `_eventols` or `_eventiv` to `_eventgenvars` as an option.

@jorpppp 
thanks. In https://github.com/JMSLab/xtevent/commit/d7ae4e6ee6bb3a31b238ac27ab48799071062b5e I reverted the changes to `_eventgenvars`'s syntax and instead of, I passed the marked sample variable from `_eventols`, `_eventiv`, etc. as an option in `_eventgenvars`. Then, I ran the do-file in the issue folder to check that this change worked.

Constantino-Carreto-Romero commented:
When running 
```
use example31, clear
xtevent y eta, panelvar(i) timevar(t) pol(z) impute(stag) window(max) diffavg 
```

There is a bug because we get a calculated left endpoint of -19, but the computation of the differences of averages only includes until coefficient -9.
